lib/record_store/zone/config/template.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+module RecordStore
+  class Zone
+    class Config
+      class TemplateContext
+        def self.build(record:, all_records:)
+          new(record: record, all_records: all_records)
+        end
+
+        def initialize(record:, all_records:)
+          @record = record
+          @all_records = all_records
+        end
+
+        def get_binding
+          binding
+        end
+
+        private
+
+        attr_reader :record, :all_records
+      end
+
+      class Template
+        class Error < StandardError; end
+        class InvalidFilepathError < Error; end
+
+        class << self
+          def from_file(filename:)
+            filepath = template_filepath_for(filename: filename)
+
+            raise InvalidFilepathError, "Could not find template '#{filename}' at location #{filepath}\n"\
+              unless File.exists?(filepath)
+
+            template_file = File.read(filepath)
+            filters_for_records_to_template = YAML.load(template_file).deep_symbolize_keys[:each_record]
+
+            new(template: ERB.new(template_file), filters_for_records_to_template: filters_for_records_to_template)
+          end
+
+          private
+
+          def template_filepath_for(filename:)
+            "#{RecordStore.templates_path}/#{filename}"
+          end
+        end
+
+        def initialize(template:, filters_for_records_to_template:)
+          @template = template
+          @filters_for_records_to_template = filters_for_records_to_template
+        end
+
+        def inject_records!(all_records:)
+          all_records
+            .select { |record| should_template?(record: record) }
+            .map { |record| template_record_for(record: record, all_records: all_records) }
+            .each do|template_record|
+              if should_inject?(template_record: template_record, all_records: all_records)
+                all_records << Record.build_from_yaml_definition(template_record[:injected_record].first)
+              end
+            end
```

Comment:
So we want to keep that `if` in the loop to avoid injecting twice the same record, for instance, if it matches two 'A' records. But we can avoid the extra block: (nitpicking though)

```suggestion
            .each do |template_record|
              next unless should_inject?(template_record: template_record, all_records: all_records)
              all_records << Record.build_from_yaml_definition(template_record[:injected_record].first)
            end
```

Reply:
I like it!